src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -284,7 +284,7 @@ JRT_LEAF(jlong, SharedRuntime::lrem(jlong y, jlong x))
 JRT_END
 
 
-#ifdef _WIN64
+#ifdef _M_ARM64
```

Comment:
Wouldn't this cause a multiple definition error for frem and drem if the \#if !defined(X86) was removed? The \#if prevents those 2 methods from being defined when on x86 so it doesn't conflict with the x86 specific definitions, without it frem and drem would be defined twice on x86

Reply:
But if we enter this block because _M_ARM64 is defined then X86 cannot be defined so line 294 is vacuously true.

Reply:
I'm not sure I follow \- The \_M\_ARM64 block is terminated by an endif before line 294, it is only there to define the global constants and does not guard the code below that line 294 does. Additionally, the code guarded by the if X86 seems to be for all non x86 platforms, not just ARM64

Reply:
Apologies somehow the endif at line 292 was invisible

Reply:
Though now I am even more confused. For x64 where are the definitions for `float_sign_mask` etc??

Reply:
x86 (And by extension x64) uses handwritten assembly to implement this, and as such simply doesn't use float\_sign\_mask and friends: https://github.com/openjdk/jdk/blob/3b7571d37812472a2152f9c8cbfd2a4abdb35016/src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp#L123

Reply:
Sorry still struggling to understand what os+cpu combinations will cause this code to be compiled. If not ARM64 and not x64 then where are the definitions for things like `float_sign_mask` used by this seemingly shared code?

Reply:
float\_sign\_mask and friends are only used by Windows ARM64, in the workaround for the Windows CRT bug (I do wish we could get rid of it though!). If looked at closely, the workaround that uses float\_sign\_mask and friends is only compiled when !X86 && \_WIN64, which means Windows ARM64. To simplify:

- All x86 platforms, including x86, use handwritten assembly, so doesn't use float\_sign\_mask and friends
- Windows ARM64 uses float\_sign\_mask and friends, as it has to use the workaround
- All other platforms simply directly call fmod without using float\_sign\_mask and friends at all, so there is no definition for those platforms

Reply:
So why does the _M_ARM64 block  not cover all of this code instead of it appearing to be used by other platforms, when it isn't?

Reply:
I'm not really sure what you mean, if you mean why the \_M\_ARM64 ifdef doesn't guard the frem and drem implementations in sharedRuntime.cpp here, it's because the Windows/ARM64 implementation with the workaround in it and the implementation for the rest of the platforms (minus x86) are squashed right next to each other. Which implementation is compiled into HotSpot is then determined by the \#ifdef \_WIN64 check on line 296. If the \_M\_ARM64 ifdef covered the frem and drem implementation here then the implementation for all non Windows/ARM64 and non x86 platforms would not be defined, and HotSpot wouldn't link at all. Well, that's if I understood you correctly, at least

It might be possible to split the Windows/ARM64 implementation specifically into another file to reduce the confusion, but I feel like that is more trouble than it's worth, and may just cause more issues if not done right. I'd prefer not to do that to avoid all those problems

Reply:
Sorry I'm doing a really bad job of pretending to be the preprocessor. :)

So the block at  line 287 defines some constants for all ARM64, even though they are only needed on Windows.

The block at line 294 provides function definitions for all CPUs other the x86. Then within that we have one version for WIN64 (which really means Windows/Aarch64) and a regular version for all other non-x86 CPUS and non-Windows Aarch64.

So I guess what I'm suggesting is that it would be clearer to use `if defined(_M_ARM64) && defined(WINDOWS)` at lines 287, 294 and 321, to be clear the special code is only for Windows/Aarch64. 

Second I think it would be clearer to just have two blocks: one for windows/aarch64 and the other for all other non-x86, rather than interleaving things the way they have been done.

Reply:
Ah, I see what you mean. There's just a small change I would make, which is to check whether defined(\_WINDOWS) and defined(AARCH64) instead. Checking for both \_M\_ARM64 and \_WINDOWS is redundant; \_M\_ARM64 is a Windows only define, so checking for \_M\_ARM64 == checking for AARCH64 && \_WINDOWS. Would you be ok with that instead?

I think(?) I get what you mean by having 2 blocks, I'll try to do that

> Sorry I'm doing a really bad job of pretending to be the preprocessor. :)

Haha, it happens

Reply:
> Ah, I see what you mean. There's just a small change I would make, which is to check whether defined(_WINDOWS) and defined(AARCH64) instead. Checking for both _M_ARM64 and _WINDOWS is redundant; _M_ARM64 is a Windows only define, so checking for _M_ARM64 == checking for AARCH64 && _WINDOWS.

Checking for `AARCH64 && _WINDOWS` is much easier to understand.

Reply:
Andrew and David, are the latest changes what you had in mind?